src/transport_picoquic.cpp

```diff
@@ -325,99 +327,103 @@ PqLoopCb(picoquic_quic_t* quic, picoquic_packet_loop_cb_enum cb_mode, void* call
-                    // Use this time to clean up streams that have been closed
-                    transport->RemoveClosedStreams();
+            if (!transport->pq_loop_prev_time) {
+                transport->pq_loop_prev_time = targ->current_time;
+            }
 
-                    if (transport->pq_loop_metrics_prev_time) {
-                        transport->EmitMetrics();
-                    }
+            if (targ->current_time - transport->pq_loop_metrics_prev_time >= kMetricsIntervalUs) {
+                // Use this time to clean up streams that have been closed
+                transport->RemoveClosedStreams();
 
-                    transport->pq_loop_metrics_prev_time = targ->current_time;
+                if (transport->pq_loop_metrics_prev_time) {
+                    transport->EmitMetrics();
                 }
 
-                if (targ->current_time - transport->pq_loop_prev_time > 100'000) {
+                transport->pq_loop_metrics_prev_time = targ->current_time;
+            }
 
-                    transport->CheckConnsForCongestion();
+            if (targ->current_time - transport->pq_loop_prev_time > 100'000) {
 
-                    transport->pq_loop_prev_time = targ->current_time;
-                }
+                transport->CheckConnsForCongestion();
 
-                // Stop loop if shutting down
-                if (transport->Status() == TransportStatus::kShutdown) {
-                    SPDLOG_LOGGER_INFO(transport->logger, "picoquic is shutting down");
+                transport->pq_loop_prev_time = targ->current_time;
+            }
 
-                    picoquic_cnx_t* close_cnx = picoquic_get_first_cnx(quic);
+            // Stop loop if done shutting down
+            if (transport->Status() == TransportStatus::kShutdown) {
+                return PICOQUIC_NO_ERROR_TERMINATE_PACKET_LOOP;
+            }
 
-                    if (close_cnx == NULL) {
-                        return PICOQUIC_NO_ERROR_TERMINATE_PACKET_LOOP;
-                    }
+            if (transport->Status() == TransportStatus::kShuttingDown) {
+                SPDLOG_LOGGER_INFO(transport->logger, "picoquic is shutting down");
 
-                    while (close_cnx != NULL) {
-                        SPDLOG_LOGGER_INFO(
-                          transport->logger, "Closing connection id {0}", reinterpret_cast<uint64_t>(close_cnx));
-                        picoquic_close(close_cnx, 0);
-                        close_cnx = picoquic_get_next_cnx(close_cnx);
-                    }
+                picoquic_cnx_t* close_cnx = picoquic_get_first_cnx(quic);
 
+                if (close_cnx == NULL) {
                     return PICOQUIC_NO_ERROR_TERMINATE_PACKET_LOOP;
                 }
 
-                break;
+                while (close_cnx != NULL) {
+                    SPDLOG_LOGGER_INFO(
+                      transport->logger, "Closing connection id {0}", reinterpret_cast<uint64_t>(close_cnx));
```

Comment:
Reinterpret cast seems strong, what type is close_cnx? Would static cast work?

Reply:
It's a pointer, so we have to reinterpret.